Multiple direction support

felix-dumit opened the pull request:
#118 

Added the ability to swipe in any direction.
Currently implemented directions are: 

```swift
public enum SwipeResultDirection:String {
    case None
    case Left
    case Right
    case Up
    case Down
    case TopLeft
    case TopRight
    case BottomLeft
    case BottomRight
}
```

It could also be extended to another custom direction by specifying a direction point (the coordinate system has its origin at the center to the cardView).

Most drastic changes are:
 -  to determine the swiped direction it compares the current drag vector to the defined direction vectors and determines which direction the drag has the larges projection
 - to determine the swipe percentage, it calculates the normalised distance between the center point of the card, and the border point of the card in the direction in question
   - Also changes had to be made to the animation code so it would be responsive to all direction drags - **this is the part I'm least familiar with in the code so I'd advice special care when reviewing these changes**

AEugene commented:
Hi @felix-dumit . I'll review your code by the end of the week. Sorry for delay.

felix-dumit commented:
@AEugene Any updates on this? 
I saw you made some pretty big changes so now this branch has conflicts. How do we want to proceed?

AEugene commented:
@felix-dumit Sorry for delay again, there were some urgent tasks. Could you update your branch with latest master? And then I'll provide review. Thank you for contribution.

felix-dumit commented:
@AEugene Just updated with the changes from `master`

AEugene commented on Pod/Classes/KolodaView/SwipeResultDirection.swift:
```diff
@@ -0,0 +1,183 @@
+    }
+    
+    func normalizedPointForSize(screenSize:CGSize) -> CGPoint {
+        let x = 2 * (self.x / screenSize.width)
```
Why do you multiple by 2 in normalisation? Also parameter called screenSize, but you're actually using bounds of DraggableCardView.

felix-dumit replied:
Ok this function name was actually misleading, I added another function that performed the normalisation correctly. 
What this one does is normalize a distance based on the screensize. It is multiplied by two because the normalized coordinate system goes from (-1, 1)

AEugene commented on Pod/Classes/KolodaView/SwipeResultDirection.swift:
```diff
@@ -0,0 +1,183 @@
+        return dotProductWith(point)/point.modulo
+    }
+    
+    func scalarProjectionPointWith(point:CGPoint) -> CGPoint {
```
Could you describe purpose of this method?

felix-dumit replied:
Ok there was actually a bug in that created point's `y` needed to be multiplied by `r` and not `y`.
But the purpose of this method is to calculate the point C in the figure below:
![untitled](https://cloud.githubusercontent.com/assets/1863251/14570245/2041951c-02f8-11e6-85f4-e5668f0e4dcd.png)

AEugene commented on Pod/Classes/KolodaView/DraggableCardView/DraggableCardView.swift:
```diff
@@ -214,9 +221,9 @@ public class DraggableCardView: UIView {
+            updateOverlayWithFinishPercent(dragPercentage, direction:dragDirection)
             //100% - for proportion
-            delegate?.card(self, wasDraggedWithFinishPercentage: min(fabs(dragDistance.x * 100 / CGRectGetWidth(frame)), 100), inDirection: dragDirection)
+            delegate?.card(self, wasDraggedWithFinishPercentage: min(fabs(100 * dragPercentage), 100), inDirection: dragDirection)
```
You calculate dragPercentage 2 times per one movement, better to create variable inside this scope and save value there.

AEugene commented:
@felix-dumit I've found a bug. It reproduces with horizontal swiping:
https://www.dropbox.com/s/fnmh686741yaane/HorizontalBug.mov?dl=0

Issue is in dragPercentage calculations. I've tried this approach:
```swift    
private var dragPercentage:CGFloat {
        // normalize dragDistance then convert project closesest direction vector
        let normalizedDragPoint = dragDistance.normalizedPointForSize(bounds.size)
        let swipePoint = normalizedDragPoint.scalarProjectionPointWith(dragDirection.point)
        
        // if point is outside rect, percentage of swipe in direction is over 100%
        if !SwipeResultDirection.boundsRect.contains(swipePoint) {
            return 1.0
        } else {
            let projectionPoint: CGPoint = directions.reduce(dragDirection.point) {
                if $0.distanceTo(normalizedDragPoint) > $1.point.distanceTo(normalizedDragPoint) {
                    return $1.point
                }
                
                return $0
            }
           
            return normalizedDragPoint.scalarProjectionWith(projectionPoint)
        }
    }
```
And it works great with horizontal and vertical swiping, but there is obvious jumping with directions set ` [.BottomLeft, .Left, .BottomRight, .Right] `

I can't merge it now, because horizontal swiping is key functionality of Koloda.

And setting currentCardIndex requires more complex logic. For example when you trying to set 
 `koloda.changeCurrentCardIndex = koloda.currentCardIndex + 1 `
All cards will disappear event though next cards is on the screen now. Lets leave this functionality now and focus on multiple directions swiping.

Also there are some syntax issues(like space after colon, etc). I started refactoring Koloda, so lets keep it clean. I suggest you to read: https://swift.org/documentation/api-design-guidelines/

Thank you for contribution. I'm waiting for updated pull request :)

felix-dumit commented:
I'll take a closer look on Monday, but regarding the issue you mentioned can you verify that the `swipePercentageMargin` set in the delegate is < 1?
Because previously the delegate method returned the necessary drag distance and now it's the percentage of distance between center point and edge of card at drag direction.

Anyway thanks for review I'll update soon

AEugene commented:
@felix-dumit It is reproducible with default threshold value. You can try it on first example(one with red background).

felix-dumit commented:
@Ok I was able to reproduce, you have to start dragging at the very center or the card either up or down then the I get the issue.
I'll work on a fix

felix-dumit commented:
@AEugene Ok I fixed the issue. It was because if the dragDistance was very small (rounded to CGPoint.zero) it would not intersect with any of the target lines so the `dragPercentage` would be `CGFloat.infinity`

I updated so that now it will return 0 if no instersections are found:
```swift
return rect.perimeterLines
            .flatMap { CGPoint.intersectionBetweenLines(targetLine, line2: $0) }
            .map { centerDistance / $0.distanceTo(.zero) }
            .minElement() ?? 0
```

felix-dumit commented:
Should be ready for re-review now

weyert commented:
This sounds great 👍

AEugene commented on Pod/Classes/KolodaView/DraggableCardView/DraggableCardView.swift:
```diff
@@ -62,7 +66,11 @@ public class DraggableCardView: UIView {
     override public var frame: CGRect {
         didSet {
-            actionMargin = delegate?.card(cardSwipeThresholdMargin: self) ?? frame.size.width / 2.0
+            if let r = delegate?.card(cardSwipeThresholdRatioMargin: self) where r != 0 {
```
could you name variable as "ratio" here?

AEugene commented on Pod/Classes/KolodaView/DraggableCardView/DraggableCardView.swift:
```diff
@@ -233,42 +242,83 @@ public class DraggableCardView: UIView {
-        return dragDistance.x > 0 ? .Right : .Left
+        //find closest direction
+        let normalizedDragPoint = dragDistance.normalizedDistanceForSize(bounds.size)
+        return directions.reduce((CGFloat.infinity, .None)) { min, dir in
```
I think this shorthands works great in Geometry file, but here variables should be as descriptive as possible.

AEugene commented:
hi @felix-dumit . Code looks great, you did good job. Thank you for contribution.
I left a few comments. Also could you discard changes in following files:
CustomOverlayView.swift
ExampleOverlayView.swift
ViewController.swift
OverlayView.xib
Main.storyboard

Your code will be merged after these changes.
Our team will provide designs for example project (Up/Down swiping overlay, etc), and then I’ll integrate it into Koloda.

Also I’ve noticed finish percentage jumping with configuration  ` [.BottomLeft, .Left, .BottomRight, .Right] ` on transition from left to bottom-left mode. But that’s minor issue.

felix-dumit commented:
@AEugene Ok just pushed those changes, I had to leave the change from `OverlayMode` to `SwipeResultDirection` in `CustomOverlayView.swift` and `ExampleOverlayView.swift`.

brsunter commented:
@felix-dumit I'm testing your branch and I'm getting some weird behaviour when using 
```
 func koloda(koloda: KolodaView, allowedDirectionsForIndex index: UInt) -> [SwipeResultDirection]{
        return [SwipeResultDirection.None]
    }
```

felix-dumit commented:
@brsunter yeah that's kind of undefined behaviour for that scenario as it doesn't make much sense to use it. If you don't want to allow any directions just return an empty array.

Maybe there could be an assert preventing your example. Or even remove the `None` option as I'm not sure it's even serving any purpose at the moment.
What do you think @augene?

brsunter commented:
I would remove the "None" case and have an empty array represent "no valid direction" so you can't do things like [.Left, .None]

brsunter commented:
I think removing `.None `would be best since it also prevents `[.None, .Down]` etc. Not sure about Optional vs` []` though

AEugene commented:
@felix-dumit @brsunter Yup, I think we should remove   `.None` and make allowedDirections optional
 ```swift
    func koloda(koloda: KolodaView, allowedDirectionsForIndex index: UInt) -> [SwipeResultDirection]?
 ```

felix-dumit commented:
@AEugene I think we don't need to make `allowedDirections` optional as users can still return an empty array. The `dragDirection` property in `DraggableCardView` will have to be made optional.

AEugene commented:
@felix-dumit Hm, maybe you're right. It will complicate the logic, because we still need to handle empty array of directions. Also we need to handle nil direction when we update overlay state. Could you implement it?

felix-dumit commented:
@AEugene Ok I just pushed that, tell me what you think.

AEugene commented on Pod/Classes/KolodaView/DraggableCardView/DraggableCardView.swift:
```diff
@@ -222,10 +222,12 @@ public class DraggableCardView: UIView {
-            delegate?.card(self, wasDraggedWithFinishPercentage: min(fabs(100 * percentage), 100), inDirection: dragDirection)
+            if let dragDirection = dragDirection {
+                let percentage = dragPercentage
+                updateOverlayWithFinishPercent(percentage, direction:dragDirection)
```
Shouldn't we handle nil direction to update overlay image?

felix-dumit replied:
Yes, you're right, I just fixed this.

AEugene commented:
@felix-dumit I merged your request. Thanks for contribution.

felix-dumit commented:
@AEugene Great!
